### PR TITLE
draft notes for our github flow workflow - ticket 508 

### PR DESCRIPTION
**Pull request by pkimber**
@mdinsmore Trying to learn GitHub Flow and document it at the same time.  What do you think so far?
@gbushell and @timitee you might be interested as well :)

### CODE REVIEW

**Comment by gbushell** [comment]
Test comment... I can automatically merge the pull request if you want me too

**Comment by gbushell** [comment]
Test comment 2... But I've just read you're notes... I don't want to automatically merge the pull request yet!!

File: source/dev-workflow.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+.. highlight:: bash
+
+GitHub Flow
+===========
+
+We are using the GitHub Flow.
+
+Here is a 5 minute outline of the process: `Understanding the GitHub Flow`_
+
```
Comment (gbushell) [inline source/dev-workflow.rst]: I'll watch this on 13/1/15...

Comment (gbushell) [inline source/dev-workflow.rst]: Test line note... what does this do?

**Comment by pkimber** [comment]
This is very impressive.  I think it will be great for working together on code etc....

**Comment by mdinsmore** [comment]
I've changed to the 508-dev-workflow branch then altered the README.rst to alter the title to Documentation... and pushed the result

**Comment by mdinsmore** [comment]
I think it's a good way to talk about changes on a specific topic.  I agree the master branch should be deployable all the time and we should use branches for major changes.  Pull requests seem like a reasonable way to be able to talk about the code and review code from each other.  

I have several questions

**Is there a way to see the what branches have activity?**
I have to confess I find it quite difficult it is to see what is going on.  For example when the master branch is selected there is no indication that there is activity on a branch. Or have I missed something (I presume I'm missing something here the guy in the video from github was so excited about it)

**Is there any way to tie in changes from other repositories?**
As far as I can see we cannot tie in changes to other repositories for example suppose we have a project for a client that requires an update to the the stock app and the pay app.  We need to create a branch on each of the three repositories involved (stock app, pay app and project) We then have to remember which repositories are involved and if there is discussion about changes it could get very fragmented.

**Is there any way to merge a commit from one branch into another branch?**
For example, if Patrick fixes something on one branch that is not going to be merged into the master for some time that change will not then be automatically available to another branch that Greg is working on.

**How do we decide that the discussion on a pull request is finished?**

**Comment by pkimber** [comment]
**Is there a way to see the what branches have activity?**

I think some of the following might help:

* There is a list of pull requests: https://github.com/pkimber/docs/pulls
* We can assign labels.  I just added a "question" label.
* We get emails when there is activity on a branch
* We can link a user by typing @mdinsmore 

**Is there any way to tie in changes from other repositories?**

Good point.  I think you are correct.  We will need a  branch on both repositories.  We could try and keep discussion on a change to one of the repositories.  This shouldn't happen too much I hope!

**Is there any way to merge a commit from one branch into another branch?**

I think there is a way... but I am not so good at using git yet.  I am sure we can learn.

**How do we decide that the discussion on a pull request is finished?**

I guess discussion on a pull request is finished when it is merged back into the master branch.  If we try not to merge our own pull requests, then when the other person is happy, they can just merge it.

**Comment by mdinsmore** [comment]
Ah ok, so you only raise a pull request once development is finished. So the work flow would be something like this:

Development Stage
* New ticket raised on CRM and allocated to Developer
* Developer - Create a branch from the master
* Developer - Do some development and testing locally and commit and push
* Developer - More development and testing locally and commit work and push
* Developer - Raises a pull request 
* Developer - Make a comment asking for review by Reviewer

Review Stage (may be several iterations)
* Reviewer - Review commits and either make comments to Developer or start merge stage
* if necessary Developer - Make changes, commit, push and add a comment requesting another review

Finalise Stage
* Reviewer  - Merges Pull Request
* Developer - Create a Release
* Developer - Deploy

**Comment by pkimber** [comment]
Yes... that is exactly how I understand it.  I like that :+1: 

* We commit and push the code during development so we have an offsite copy of the work done so far.
* We make a pull request when we want the work reviewed.

The video helped me understand that we could do big pieces of work in a pull request.  Before listening to the video I thought they were only for little bits of work.

If you are happy to give this a go, then Greg and I will look at paying for GitHub so we can have private repositories.

**Comment by mdinsmore** [comment]
Have you looked at bitbucket's system for this?  It has a way of viewing branches and has a pull request mechanism that allows comments.

**Comment by pkimber** [comment]
No.  Shall we try that one next time we do a fix or feature request?

**Comment by mdinsmore** [comment]
I think that makes sense

**Pull request merged**